gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/MysqlDagStateStoreWithDagNodes.java

```diff
@@ -126,18 +126,32 @@ public MysqlDagStateStoreWithDagNodes(Config config, Map<URI, TopologySpec> topo
   }
 
   @Override
-  public void writeCheckpoint(Dag<JobExecutionPlan> dag)
-      throws IOException {
-    DagManager.DagId dagId = DagManagerUtils.generateDagId(dag);
-    boolean newDag = false;
-    for (Dag.DagNode<JobExecutionPlan> dagNode : dag.getNodes()) {
-      if (updateDagNode(dagId, dagNode) == 1) {
-        newDag = true;
+  public void writeCheckpoint(Dag<JobExecutionPlan> dag) throws IOException {
+    String dagId = DagManagerUtils.generateDagId(dag).toString();
+    dbStatementExecutor.withPreparedStatement(String.format(INSERT_DAG_NODE_STATEMENT, tableName), insertStatement -> {
+      int dagSize = dag.getNodes().size();
+      Object[][] data = new Object[dagSize][3];
+
+      for (int i=0; i<dagSize; i++) {
+        Dag.DagNode<JobExecutionPlan> dagNode = dag.getNodes().get(i);
+        data[i][0] = dagNode.getValue().getId().toString();
+        data[i][1] = dagId;
+        data[i][2] =this.serDe.serialize(Collections.singletonList(dagNode.getValue()));
       }
-    }
-    if (newDag) {
-      this.totalDagCount.inc();
-    }
+
+      for (Object[] row : data) {
```

Comment:
why iterate through to populate data when the only thing done w/ it is to iterate over it here?  instead, just iterate through the `DagNode`s and populate `insertStatement` directly.